File: app/src/main/java/ua/com/foxminded/locationtrackera/mvi/MviViewModel.java
package ua.com.foxminded.locationtrackera.mvi;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleEventObserver;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.disposables.Disposable;

public abstract class MviViewModel<S, A> extends ViewModel implements FragmentContract.ViewModel<S, A>, LifecycleEventObserver {

    private final CompositeDisposable onStopDisposable = new CompositeDisposable();
    private final CompositeDisposable onDestroyDisposable = new CompositeDisposable();
    private final MutableLiveData<S> stateHolder = new MutableLiveData<>();
    private final MutableLiveData<A> actionHolder = new MutableLiveData<>();

    @Override
    public MutableLiveData<S> getStateObservable() {
        return stateHolder;
    }

    @Override
    public MutableLiveData<A> getEffectObservable() {
        return actionHolder;
    }

    @CallSuper
    @Override
    public void onStateChanged(@NonNull LifecycleOwner source, @NonNull Lifecycle.Event event) {
        switch (event) {
            case ON_STOP:
                onStopDisposable.clear();
            case ON_DESTROY:
                onDestroyDisposable.clear();
        }
    }

    protected void setState(S state) {
        stateHolder.setValue(state);
    }

    protected void postState(S state) {
        stateHolder.postValue(state);
    }

    protected void setAction(A action) {
        actionHolder.setValue(action);
    }

    protected void postAction(A action) {
        actionHolder.postValue(action);
    }

    protected void addTillDestroy(Disposable... disposables) {
        onDestroyDisposable.addAll(disposables);
    }

    @CallSuper
    @Override
    protected void onCleared() {
        onDestroyDisposable.dispose();
        onStopDisposable.dispose();
    }
}

File: mvi/src/main/java/ua/com/foxminded/locationtrackera/mvi/MviViewModel.kt
package ua.com.foxminded.locationtrackera.mvi

import androidx.annotation.CallSuper
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.disposables.Disposable
import ua.com.foxminded.locationtrackera.mvi.fragments.FragmentContract

abstract class MviViewModel<S, A> : ViewModel(), FragmentContract.ViewModel<S, A> {

    private val onStopDisposable = CompositeDisposable()
    private val onDestroyDisposable = CompositeDisposable()
    private val stateHolder = MutableLiveData<S>()
    private val actionHolder = MutableLiveData<A>()

    override fun getStateObservable(): MutableLiveData<S> = stateHolder

    override fun getEffectObservable(): MutableLiveData<A> = actionHolder

    @CallSuper
    override fun onStateChanged(event: Lifecycle.Event) {
        when (event) {
            Lifecycle.Event.ON_STOP -> {
                onStopDisposable.clear()
                onDestroyDisposable.clear()
            }
            Lifecycle.Event.ON_DESTROY -> onDestroyDisposable.clear()
            else -> return
        }
    }

    protected fun setState(state: S) {
        stateHolder.value = state
    }

    protected fun setAction(action: A) {
        actionHolder.value = action
    }

    protected fun addTillDestroy(vararg disposables: Disposable) {
        onDestroyDisposable.addAll(*disposables)
    }

    @CallSuper
    override fun onCleared() {
        onDestroyDisposable.dispose()
        onStopDisposable.dispose()
    }
}
